### ISSUE TITLE
Dynamically adding fields in the schema doesn't permit to add fields as usual

### ISSUE THREAD

**Issue by pmaoui** [Automattic/mongoose#1867]
Hi,

When I dynamically add a field to a schema and instantiate it to a model, mongoose is not aware of any change made on this new field when I call save. 

```
var schema = new mongoose.Schema({
  name: String,
  firstname: String
});

var person = db.model('persons', schema);
var a = new person();
a.name = "John";
a.firstname = "Doh";

a.save(function(err, savedGuy) {

  schema.add({
    field1: String
  });
  // schema is now updated with the field1
  var personV2 = db.model('persons', schema);

  personV2.findById(savedGuy['_id'], function(err, john) {

    // won't work
    john.field1 = 'Custom field 1 value';
    // will work
    john.set ('field1', 'Custom field 1 value');

    });
});
```

I create dynamic but persistent schemas in my project. I can't fill any data until I restart node. It looks like a bug for me.
The purpose of schema.add function should be to avoid the use of set() which seems useful only for arrays properties that can't be detected by mongoose. 

I need right now to intensively manipulating objects which is not pretty at all. So my questions are : is this an issue ? If so, is this possible to think about a fix ?

Best regards and thank you for your work !

**Comment by tomasdev** [comment]
:+1: @lijinma :heart:

**Comment by fishmacs** [comment]
Seems disable mongoose strict conflict with discriminator?

Error: Discriminator options are not customizable (except toJSON & toObject)
...

I want to add some fields on fly if some plugins loaded, use the original model, is this possible?

**Comment by vkarpov15** [comment]
@fishmacs can you clarify? I don't understand your issue.

**Comment by fishmacs** [comment]
@vkarpov15 thanks!

Say I have some discriminated **User** models:

    var UserSchema = new Schema({
        username: String;
        password: String;
    });

    var AdminSchema = new Schema ({
       email: String
    });

    var User = mongoose.model('User', UserSchema);
    User.discriminator('Admin', AdminSchema);

Then there is a **Course** model in another module, this Course model tries to add more fields to User:

    var CourseSchema = new Schema({
       name: String;
       teacher: String;
       ......
    });

    var Course = mongoose.model('Course', CourseSchema);

    mongoose.model('User').schema.add({
      'courses': [{type: ObjectId, ref: 'Course'}]
    });

I hope *user.courses* can represent the courses selected by the user, but user.courses is always **undefined**.

I tried to add *{strict: false}* to UserSchema(in fact, I'm not sure what **strict** means, if it is what I want), but seems it conflict with **discriminator**:

    Error: Discriminator options are not customizable (except toJSON & toObject)
       at /myproj/courseapp/node_modules/mongoose/lib/model.js:783:15
       at /myproj/courseapp/node_modules/mongoose/lib/model.js:786:7
       at Function.discriminator (/myproj/courseapp/node_modules/mongoose/lib/model.js:797:5)
       at Object.<anonymous> (/myproj/courseapp/src/models/user.js:77:6)

I also tried **Mixed** type, but no help.

**Comment by vkarpov15** [comment]
1) Adding paths to the model's schema after the model has been compiled is unsupported. Please don't do that.
2) [strict mode](http://mongoosejs.com/docs/guide.html#strict)
3) When you use `discriminator()`, you need to pass the same options you passed to `CourseSchema` and `UserSchema`. You can change the `toJSON` and `toObject` schema options, but not `strict`. Admittedly this is somewhat cumbersome and will likely change in the future (#3414). For more information on discriminators, check out [my blog post on discriminators](http://thecodebarbarian.com/2015/07/24/guide-to-mongoose-discriminators).

**Comment by fishmacs** [comment]
@vkarpov15 Thanks! 
And, great post on discriminator! :+1:

**Comment by nosliwsirhc** [comment]
I added a "subscription" property but Node isn't returning any of the new data. I used Robo 3T to login to MongoDB and manually entered the data to the existing few records but when I run a query it shows every property except for "subscription". I also added a new record after changing the schema/reloading Node to see if that made a difference but still the same issue.

`const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const serviceProviderSchema = new Schema({
    name: String,
    logoFile: {
        fieldname: String,
        originalname: String,
        encoding: String,
        mimetype: String,
        size: Number,
        bucket: String,
        key: String,
        acl: String,
        contentType: String,
        contentDisposition: String,
        storageClass: String,
        serverSideEncryption: String,
        metadata: { fieldName: String },
        location: String,
        etag: String
    },
    description: String,
    address: {
        street: String,
        unit: String,
        city: String,
        province: String,
        postalCode: String
    },
    contact: {
        mainPhone: String,
        fax: String,
        placementPhone: String,
        placementPhoneExtension: String,
        mainEmail: String,
        placementEmail: String
    },
    perDiems: [{
        label: String,
        price: Number,
        description: String
    }],
    subscription: {
        type: String,
        subscriberSince: String,
        renewals: [{
            begin: String,
            expiry: String,
            transactionId: String
        }]
    },
    residences: [{
        _id: {type: Schema.Types.ObjectId, ref: 'residence'},
        name: String
    }],
    users: [{
        _id: {type: Schema.Types.ObjectId, ref: 'user'},
        role: String
    }]

});

const ServiceProvider = mongoose.model('serviceProvider', serviceProviderSchema);
module.exports = ServiceProvider;`

**Comment by vkarpov15** [comment]
@nosliwsirhc Read this: http://mongoosejs.com/docs/faq.html#type-key

TLDR, do this instead:

```javascript
    subscription: {
        type: { type: String },
```

**Comment by nosliwsirhc** [comment]
Awesome, didn’t know that it was a reserved keyword. Made the adjustment, for me I changed it from “type” to “subscriptionType”, now it works!

**Comment by hassantauqeer** [comment]
How to push in schema? like this:
````
const supported = require('../consts');

var XSchema = new mongoose.Schema(
  return () => {for(let items in supported) {
    supported[items]: {
        default: 'x',
        type: string
      }
  }}
);
````

**Comment by vkarpov15** [comment]
@hassantauqeer I don't understand your example, it's pretty far off from syntactically valid JavaScript. Maybe you're looking for eachPath? https://mongoosejs.com/docs/api.html#schema_Schema-eachPath . If not, please clarify what you mean.

**Comment by kaleemmolani** [comment]
> @vkarpov15 thanks!
> 
> Say I have some discriminated **User** models:
> 
> ```
> var UserSchema = new Schema({
>     username: String;
>     password: String;
> });
> 
> var AdminSchema = new Schema ({
>    email: String
> });
> 
> var User = mongoose.model('User', UserSchema);
> User.discriminator('Admin', AdminSchema);
> ```
> 
> Then there is a **Course** model in another module, this Course model tries to add more fields to User:
> 
> ```
> var CourseSchema = new Schema({
>    name: String;
>    teacher: String;
>    ......
> });
> 
> var Course = mongoose.model('Course', CourseSchema);
> 
> mongoose.model('User').schema.add({
>   'courses': [{type: ObjectId, ref: 'Course'}]
> });
> ```
> 
> I hope _user.courses_ can represent the courses selected by the user, but user.courses is always **undefined**.
> 
> I tried to add _{strict: false}_ to UserSchema(in fact, I'm not sure what **strict** means, if it is what I want), but seems it conflict with **discriminator**:
> 
> ```
> Error: Discriminator options are not customizable (except toJSON & toObject)
>    at /myproj/courseapp/node_modules/mongoose/lib/model.js:783:15
>    at /myproj/courseapp/node_modules/mongoose/lib/model.js:786:7
>    at Function.discriminator (/myproj/courseapp/node_modules/mongoose/lib/model.js:797:5)
>    at Object.<anonymous> (/myproj/courseapp/src/models/user.js:77:6)
> ```
> 
> I also tried **Mixed** type, but no help.

I tried this method and its work for me without adding {strict:false}
